### DISCUSSION TITLE
Dynamic Settings

### DISCUSSION THREAD

**Discussion by InakiRaba91** [fastapi/fastapi#11169] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from functools import lru_cache
from typing import Annotated

from fastapi import Depends, FastAPI

from pydantic_settings import BaseSettings
from pydantic_settings import BaseSettings, SettingsConfigDict


class Settings(BaseSettings):
    app_name: str

    model_config = SettingsConfigDict(env_file=".env")

app = FastAPI()

def get_settings():
    return Settings()


@app.get("/info")
def info(settings: Annotated[Settings, Depends(get_settings)]):
    return {"app_name": settings.app_name}
```

### Description

I'm trying to have a setting that I can update on the fly, with my endpoint behaving accordingly. So for instance, I'd like to spin up the server with app_name="Initial API". I've added this to the .env file

```
APP_NAME="Initial API"
```

After running the server, I update the .env file to

```
APP_NAME="Updated API"
```

and the endpoint returns the new value, so I'm getting the desired behavior.

What I'm wondering is if this is the best way to achieve this, since it implies having to read the .env file every time the endpoint is hit. It's fine for my use case, but I was just curious if there's a better approach.

Thanks a lot in advance!

### Operating System

Linux

### FastAPI Version

0.100.1

### Pydantic Version

2.4.2

### Python Version

3.11.3

**Comment by n8sty** [comment]
You can do something similar but instead of modify the `.env` file directly, load the settings into `app.state` using the application lifespan and then modify the configuration object there as a result of making requests to your application. Something like:

```
class State(TypedDict):
    config: dict[str, str]

@asynccontextmanager
async def lifespan(_: FastAPI) -> AsyncGenerator[State, None]:
    async with AsyncExitStack() as stack:
        yield {
            "config": {},
        }

app = FastAPI(lifespan=lifespan, title="Testaroo")

@app.get("/api/setconfig{key}")
async def setconfig(key: str, val: str, request: Request):
    request.state.config[key] = val
    return request.state.configthing

@app.get("/api/getconfig")
async def getconfig(request: Request):
    return request.state.config
```

**Comment by irabanillo91** [comment, marked as answer]
That works great, thanks for the suggestion!

I ended up with something a bit simpler but that should do the trick for my use case, I reckon


```python
from typing import Any
from fastapi import FastAPI
from pydantic import BaseModel


class Config(BaseModel):
    traffic: Any = None
    version: Any = None

config = Config()

app = FastAPI()

@app.get("/api/setconfig{key}")
async def setconfig(key: str, val: str):
    setattr(config, key, val)
    return config

@app.get("/api/getconfig")
async def getconfig():
    return config
```

**Reply by YuriiMotov** [reply]
Just remember, that it will only work well if you run one instance of your app (one worker).
If you are planning to run multiple workers, you can store config in Redis
